MT 20.next : Enhancements and fixes

FJBDev opened the pull request:
- 3D Map layers : 
   * Removed the deprecated or unavailable layers (**ID_USDA_NAIP & ID_USDA_NAIP_USGS & MS Virtual Earth Aerial**).
   * Added the layers **USGS NAIP PLUS, USGS Topo Base Map and USGS Topo Base Map Large**.
   * Fixed an issue where the **Atmosphere and Scale layers** would not appear when using MyTourbook in French.

(see commits [abb5f2c](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/abb5f2c88b017393dd9df6e95685d814ff17b395),  [7716589](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/771658934c7a20cdcd1828637135b25cd32bc379) & [396a402](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/396a4020a63040f0512066d75f2ffc5644cce736))

_**Example of 3D map with the USGS Topo Base Map layer**_
<img src="https://user-images.githubusercontent.com/25235665/76183710-fdd3fc00-618e-11ea-9211-96c52835a770.png" width="70%" height="70%">

<br />

- 3D Map colors: 
   * Fixed the pace ranges from seconds to minutes for better readability

(see commits [7ff5b37](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/7ff5b3755d7ee0c7218c04b2ea871fedda242183) & [7c2d298](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/7c2d2988d507947dd32feee8f52970a6db531113))

**20.3.0**
<img src="https://user-images.githubusercontent.com/25235665/77802980-b4febd00-7041-11ea-90bf-ec2a3a506361.png" width="70%" height="70%">

**20.next.0**
<img src="https://user-images.githubusercontent.com/25235665/77803008-c2b44280-7041-11ea-8bf4-b926e3f7ff19.png" width="70%" height="70%">


<br />

- 3D Map : 
   * If the data series are not available, we disable the actions

(see commit [468d8e1](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/468d8e186ed2766ea8b1caf0ac8bef8a82a33ed6))

![image](https://user-images.githubusercontent.com/25235665/77852480-28b4dd00-719c-11ea-9c62-dcee2894acb1.png)


<br />


- Statistics View : The number of years to be displayed in the graph is now dynamically filled and updated based on the number of available years
(see commit [3998b4b](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/3998b4be945c37d153d2bdb8b07fcec9151fc75e))
<br />

**20.3.0**
<img src="https://user-images.githubusercontent.com/25235665/80177797-f3908480-85b9-11ea-9481-a9f6e3f6ea9c.png" width="40%" height="40%">

**20.next.0**
<img src="https://user-images.githubusercontent.com/25235665/79075976-888aa800-7cb3-11ea-81c6-bc39472c8749.png" width="40%" height="40%">

- TCX Import bugfix : Fix for TCX files where the distance is reset at each new lap which caused the distance data and display to be wrong. Reported by [Greg Barton](https://sourceforge.net/p/mytourbook/discussion/622811/thread/4419985b92/)
(see commits [e37d119](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/e37d1193e1788423e63b0f5833df1d6156c73dfc) & [4d97703](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/4d97703f29bfc9647c6f336dd9a4cc6530fd337d))
<br />

**20.3.0**
<img src="https://user-images.githubusercontent.com/25235665/80176925-56cce780-85b7-11ea-81f5-a78b6cf06b80.png" width="85%" height="85%">

**20.next.0**
<img src="https://user-images.githubusercontent.com/25235665/80176599-57b14980-85b6-11ea-8c02-d89510e2f8e3.png" width="85%" height="85%">

<br />

- GPX/TCX Export:  Fixed and updated the GPX/TCX export templates as the file validation was
failing against the schemas.
(see commit [590e1d6](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/590e1d6690c8f36732e48535138ec0fc3d5fb6a4))
<br />

**20.3.0**
GPX Validation failures
<img src="https://user-images.githubusercontent.com/25235665/80989111-b4e89e80-8df1-11ea-95cf-aa8aa99679c6.png" width="85%" height="85%">
TCX Validation failures
<img src="https://user-images.githubusercontent.com/25235665/80989227-e3667980-8df1-11ea-8695-4f2669c0a22b.png" width="85%" height="85%">

**20.next.0**
GPX Validation success
<img src="https://user-images.githubusercontent.com/25235665/80989300-fe38ee00-8df1-11ea-94aa-26ac211418c9.png" width="85%" height="85%">
TCX Validation success
<img src="https://user-images.githubusercontent.com/25235665/80989329-098c1980-8df2-11ea-8c96-e353c31000e0.png" width="85%" height="85%">

<br />

- Updating the Garmin FIT SDK from 20.90 to **21.30** (see commit [c081fcc](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/c081fcc117c7711905a217f4987e0382b0fd0e68) & [277bd33](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/277bd3385351c7ece9c9dcb95616d24cbe976965))

<br />

- Removed the below unused jar libraries and saved around **1MB total**.
Removed unused exported packages. (see commits [6663f11](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/6663f11ed0ebf6682d7565e20fc99261e800179a) & [3e7d74d](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/3e7d74d224ee325bb925e0e4dc38759701132d82) & [c081fcc](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/c081fcc117c7711905a217f4987e0382b0fd0e68) & [c904b85](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/c904b85daba5e8361d7ff2df71126c59f00c5b39))
   - [~~xstream~~](https://mvnrepository.com/artifact/com.thoughtworks.xstream/xstream) (607 KB). Introduced in this [commit](https://github.com/wolfgang-ch/mytourbook/commit/365c2c044a9fa990f6c76b35df646e2af6013e0b), stopped being used in this [commit](https://github.com/wolfgang-ch/mytourbook/commit/2247cf348eb0028698c8bf18ac2ac13812440ce9#diff-e31e7d2c6e158d4dc47ac29ec2b9fbdf)
   - [~~xpp3~~](https://mvnrepository.com/artifact/org.ogce/xpp3) (362 KB). Introduced in this [commit](https://github.com/wolfgang-ch/mytourbook/commit/365c2c044a9fa990f6c76b35df646e2af6013e0b), stopped being used in this [commit](https://github.com/wolfgang-ch/mytourbook/commit/2247cf348eb0028698c8bf18ac2ac13812440ce9#diff-e31e7d2c6e158d4dc47ac29ec2b9fbdf). Note : Xpp3 is a dependency of XStream that was also removed
   - [~~Avalon Framework API~~](https://mvnrepository.com/artifact/org.apache.avalon.framework/avalon-framework-api) (32.4 KB). Not used in the [Apache XML Graphics 2.4 sources](http://apache.mirrors.hoobly.com/xmlgraphics/fop/source/fop-2.4-src.zip) as the Avalon project is [discontinued](https://avalon.apache.org/closed.html) since 2004

<br />

- Removed **442** warnings (Reduced the number of warnings from **1196** to **754**) (see commit [0a9e66e](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/0a9e66e94936481ec89a9ec2948400fe3c17d13e) & [2ae337a](https://github.com/wolfgang-ch/mytourbook/pull/179/commits/2ae337ad0c76cd159689dc49cd738a7f48a43eda))

<br />

 - French translation improvements :fr:

wolfgang-ch commented on bundles/net.tourbook.device/src/net/tourbook/device/hac4linux/tour/HAC4LinuxDeviceReader.java:
```diff
@@ -45,667 +45,701 @@
+      if (line.equals("[FRIENDS]")) {
+//$NON-NLS-0$
+         m_section = Section.SECTION_FRIENDS;
+         return;
+      }
+      if (line.equals("[PERSON]")) {
+//$NON-NLS-0$
+         m_section = Section.SECTION_PERSON;
+         return;
+      }
+      if (line.equals("[STATISTICS]")) {
+//$NON-NLS-0$
+         m_section = Section.SECTION_STATISTICS;
+         return;
+      }
+      if (line.equals("[SETTINGS]")) {
+//$NON-NLS-0$
+         m_section = Section.SECTION_SETTINGS;
+         return;
+      }
+      if (line.equals("[POLAR-EXTENSION]")) {
+//$NON-NLS-0$
+         m_section = Section.SECTION_POLAREXTS;
+         return;
+      }
+      if (line.equals("[COACH-PARAMETER]")) {
+//$NON-NLS-0$
+         m_section = Section.SECTION_COACH;
+         return;
+      }
+      if (line.equals("[TOUR-DATA]")) {
+//$NON-NLS-0$
+         m_section = Section.SECTION_TOURDATA;
+         return;
+      }
+      if (line.equals("[MARKING-DATA]")) {
+//$NON-NLS-0$
+         m_section = Section.SECTION_MARKS;
+         return;
+      }
```
This formatting looks very awful compared with the previous formatting. It would have been better to use the formatting off/on tags for manual formatting

FJBDev replied:
Sometimes Eclipse does weird thing with the formatting. I will revert to the original formatting if you prefer

FJBDev replied:
Fixed

FJBDev commented:
Ok, I see your comments and I will address  them soon.

wolfgang-ch commented on bundles/net.tourbook.web/src/net/tourbook/web/WebContentServer.java:
```diff
@@ -419,18 +418,17 @@ private static void handle_404(final HttpExchange httpExchange, final File file,
          httpExchange.sendResponseHeaders(200, 0);
 
-         fs = new FileInputStream(file);
+         ;
```
refactoring anomaly?

FJBDev replied:
Good catch, fixed

FJBDev replied:
Good catch. Fixed

FJBDev commented on bundles/net.tourbook.export/src/net/tourbook/export/DialogExportTour.java:
```diff
@@ -1342,7 +1341,10 @@ private void doExport_21_Creator(final VelocityContext vcContext) {
          pluginMicroVersion = Integer.toString(version.getMicro());
-         pluginQualifierVersion = version.getQualifier();
+         final String versionQualifier = version.getQualifier();
+         if (StringUtils.isNumeric(versionQualifier)) {
```
That's why I am testing if it's a numeric because the Garmin XSD schema expects an "unsignedshort"
<xsd:element name="BuildMinor" type="xsd:unsignedShort" minOccurs="0"/>
[Source](https://www8.garmin.com/xmlschemas/TrainingCenterDatabasev2.xsd)

FJBDev commented:
>  updated jar files only when there was a be



> 
> 
> Finished this PR review
> 
> This PR contains a lot of refactoring and updated jar files, this can be very dangerous when not having unit tests, I updated jar files only when there was a benefit.

That's why I asked you in 2018 why there are no unit tests in MT :-)
If you think it's too dangerous, let me know if you want me to revert to the "old" jar libraries.

FJBDev commented on bundles/net.tourbook/src/net/tourbook/preferences/PrefPageTags.java:
```diff
@@ -310,8 +310,9 @@ public void propertyChange(final PropertyChangeEvent event) {
 
-               _tagViewer.getTree().setLinesVisible(
-                     getPreferenceStore().getBoolean(ITourbookPreferences.VIEW_LAYOUT_DISPLAY_LINES));
+               _tagViewer.getTree()
+                     .setLinesVisible(
+                           getPreferenceStore().getBoolean(ITourbookPreferences.VIEW_LAYOUT_DISPLAY_LINES));
```
Yes but for some reason Eclipse sometimes "forces" to have multiple lines. I will fix manually.

FJBDev replied:
Fixed

FJBDev commented on bundles/net.tourbook/src/de/byteholder/geoclipse/map/LocalResponseCache.java:
```diff
@@ -68,9 +68,6 @@ private static boolean isUpdateAvailable(final URI remoteUri, final File localFi
 		URLConnection conn;
 		try {
 			conn = remoteUri.toURL().openConnection();
-		} catch (final MalformedURLException ex) {
-			ex.printStackTrace();
-			return false;
```
Because it's already caught in the IOException.
See Eclipse message 
![image](https://user-images.githubusercontent.com/25235665/81329571-0300da80-905c-11ea-9a79-814d4b856c81.png)

FJBDev commented on bundles/net.tourbook/src/de/byteholder/geoclipse/map/Tile.java:
```diff
@@ -1096,11 +1098,13 @@ public void setUrl(final String tileUrl) {
-            : _mapImage.isDisposed() ? //
-                  false
-                  : true;
+      /*
+       * final boolean isImageOK = _mapImage == null ? //
+       * false
+       * : _mapImage.isDisposed() ? //
+       * false
+       * : true;
+       */
```
Yes, i don't know if it's possible to tell Eclipse to use // when using the "Add Block Comments" feature in the Source menu

FJBDev replied:
Fixed

wolfgang-ch replied:
With Ctrl + 7 you can easily comment/uncomment all selected lines with // (this also works in VSCode which I now use for JS development) however when you comment a whole method then Eclipse do sometimes move the commented method in the automated save actions.

FJBDev commented on bundles/net.tourbook/src/net/tourbook/map3/ui/SlideoutMap3LayerTooltip.java:
```diff
@@ -50,8 +50,6 @@
 	private ViewerRow					_sensitiveRowArea;
 	private ViewerRow					_viewerRow;
 
-	private IToolProvider				_defaultToolProvider	= new ToolProvider();
-
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/preferences/PrefPage_TEMPLATE_With_Fields.java:
```diff
@@ -27,72 +26,72 @@
+   /*
+    * UI controls
+    */
+   //private PixelConverter		_pc;
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/preferences/PrefPage_TEMPLATE_Without_Fields.java:
```diff
@@ -30,79 +26,70 @@
 
 public class PrefPage_TEMPLATE_Without_Fields extends PreferencePage implements IWorkbenchPreferencePage {
 
-	public static final String	ID			= "net.tourbook.preferences.PrefPageGeneralExternalProgramsID"; //$NON-NLS-1$
-
-	private IPreferenceStore	_prefStore	= TourbookPlugin.getPrefStore();
-
-	/*
-	 * UI controls
-	 */
-	private PixelConverter		_pc;
-
-	@Override
-	protected Control createContents(final Composite parent) {
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tour/printing/ActionPrint.java:
```diff
@@ -44,7 +43,8 @@ public ActionPrintTour(final PrintTourExtension printTourExtension) {
 
-		ActionPrintTour(final String visibleName, final String fileExtension) {}
+      @SuppressWarnings("unused")
+      ActionPrintTour(final String visibleName, final String fileExtension) {}
```
I use Ctrl+Shift+G to find references in Eclipes (https://www.codetrails.com/blog/shortcuts-searching-eclipse/)

wolfgang-ch replied:
I also use this shortcut to find references but when you want to find a constructor with 0 arguments then you must create this constructor to find it with Ctrl+Shift+G

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/DialogTemplate.java:
```diff
@@ -32,10 +29,7 @@
 
-   private static final String    ID         = "net.tourbook.ui.DialogTemplate"; //$NON-NLS-1$
-
-   private final IDialogSettings  _state     = TourbookPlugin.getState(ID);
-   private final IPreferenceStore _prefStore = TourbookPlugin.getPrefStore();
```
Fixed

FJBDev commented on bundles/net.tourbook.printing/src/net/tourbook/printing/PrintTourPDF.java:
```diff
@@ -114,25 +114,13 @@ public void printPDF(final IXmlSerializable object, final PrintSettings printSet
             // setup xml input source
             final String xml = object.toXml();
 
-//				  // debug logging
-//				  System.err.println("--------------------------------------------------------");
-//				  System.err.println(object.toXml());
-//				  System.err.println("--------------------------------------------------------");
-//				  XStream xStream = new XStream();
-//				  try {
-//				  	FileUtils.writeStringToFile(new File("/home/jkl/tourdata_xs.xml"),
-//				  	xStream.toXML(object));
-//				  } catch (IOException e) {
-//				  	e.printStackTrace();
-//				  }
-
```
Fixed

FJBDev commented:
> 
> 
> @wolfgang-ch Ok, I see your comments and I will address them soon.

I have addressed all of them [here](https://github.com/wolfgang-ch/mytourbook/pull/184)

wolfgang-ch commented:
> That's why I asked you in 2018 why there are no unit tests in MT :-)
> As I told you at the time, I always work with Unit tests as they reduce the number of bugs, implement code coverage and once they are set up, they're automatic.
> I have a [branch ](https://github.com/FJBDev/mytourbook/tree/suunto9-unittests)where I put my Suunto 9 unit tests that I run before each new release to make sure nothing is "broken" and that there is no regresstion.
> If you think it's too dangerous, let me know if you want me to revert to the "old" jar libraries.
> 
> > I updated jar files only when there was a benefit.

We keep it now as it is and time will tell if it was good or not. When I use a new jar or update a jar then I always test the new feature which is using this jar(s).

> There is always a benefit to update jar files as typically they have bugfixes, improvements...etc
> bug fixes : Example [Updating Joda Time from 2.9.4 to 2.10.5](https://www.joda.org/joda-time/changes-report.html#a2.10.5)

Not always, one time a newer .fit jar file introduced a new bug

FJBDev commented:
> We keep it now as it is and time will tell if it was good or not. When I use a new jar or update a jar then I always test the new feature which is using this jar(s).

OK, sounds good.

> Not always, one time a newer .fit jar file introduced a new bug

I should have mentioned that after updating the fit SDK, I tested importing fit files.
Also, for all the other libraries, I did some testing as well on top of using this branch of MTB myself for weeks to see if I find more bugs when using it.

FJBDev commented on bundles/net.tourbook.common/src/net/tourbook/common/Hack.java:
```diff
@@ -23,62 +23,50 @@
-			// ignore
-		}
-	}
+      } catch (final NoSuchMethodException | IllegalArgumentException | IllegalAccessException | InvocationTargetException e) {
```
@wolfgang-ch That was a nice feature introduced in Java 7 that allows to catch multiple exceptions in 1 catch only
[Source](https://docs.oracle.com/javase/7/docs/technotes/guides/language/catch-multiple.html)
